### DISCUSSION TITLE
Global MDX components

### DISCUSSION THREAD

**Discussion by lpalazzi** [remix-run/remix#8638] [Q&A]
Is there a way to define a [components mapping object](https://mdxjs.com/docs/using-mdx/#components) that is then applied to all [MDX routes](https://remix.run/docs/en/main/guides/mdx)? The only way to apply components to an MDX module is to pass it as a prop, for example:

```tsx
import MDXComponent from './component.mdx';
import { P } from './components';

export default <MDXComponent components={{p: P}} />
```

However, if I want to use an MDX file as a route, I'm not actually calling the component anywhere so I can't pass any props.

Is there a way to do this globally?

**Comment by huijiewei** [comment]
use @mdx-js/react

layout code:

```tsx
<MDXProvider disableParentContext components={components}>
  <Suspense fallback={null}>
    <Outlet />
  </Suspense>
</MDXProvider>
```

**Reply by lpalazzi** [reply]
I've tried this, doesn't seem to work.

Here is my example:

/app/routes/posts.tsx
```tsx
import { Outlet } from '@remix-run/react';
import { MDXProvider } from '@mdx-js/react';
import { Suspense } from 'react';

export default function Index() {
  return (
    <div className='mx-auto w-full max-w-6xl px-4 sm:px-6 lg:px-8'>
      <MDXProvider
        disableParentContext
        components={{
          p: (props) => <p {...props} className='text-red-900' />,
          h1: (props) => <h1 {...props} className='text-3xl text-blue-900' />,
          h2: (props) => <h2 {...props} className='text-xl text-green-900' />,
        }}
      >
        <Suspense fallback={null}>
          <Outlet />
        </Suspense>
      </MDXProvider>
    </div>
  );
}
```

/app/routes/posts.test.mdx
```mdx
# H1 Title

## H2 Title

This is a paragraph
```

This is what is shown at the /posts/test route:

![image](https://github.com/remix-run/remix/assets/33534854/eaa0dd25-d3a5-400a-8f62-3675b6583301)

Clearly the component mapping isn't being applied as expected. But when I import the file as a module and apply the same component mapping it works:

```ts
import Test from './posts.test.mdx';

export default function Index() {
  return (
    <div className='mx-auto w-full max-w-6xl px-4 sm:px-6 lg:px-8'>
      <Test
        components={{
          p: (props) => <p {...props} className='text-red-900' />,
          h1: (props) => <h1 {...props} className='text-3xl text-blue-900' />,
          h2: (props) => <h2 {...props} className='text-xl text-green-900' />,
        }}
      />
    </div>
  );
}
```

![image](https://github.com/remix-run/remix/assets/33534854/e88bc56a-a461-4607-a96c-ed028967a46a)

**Reply by huijiewei** [reply]
Have you set the mdx plugin's configuration providerImportSource to @mdx-js/react

**Reply by lpalazzi** [reply]
No the [remix docs for MDX](https://remix.run/docs/en/main/guides/mdx) don't say anything about this and not sure where to find that config

**Reply by huijiewei** [reply]
sorry. I use remix + vite, the mdx plugin can be used alone, have more flexibility.

**Comment by Sridatta19** [comment]
All that you need to do is making this change in `vite.config.ts` for component substitution to work in remix+mdx+vite
```
mdx({
    providerImportSource: '@mdx-js/react',
}),
```

Now in blog layout you can simply pass `<MDXProvider components={components}>`

```
const components = {
	h1: H1,
	code: Code,
	link: Link,
}
```
